### PR TITLE
fix: execute JS function via blink in renderer process

### PR DESCRIPTION
**Pull request by zcbenz**
#### Description of Change

Close https://github.com/electron/electron/issues/21288.

When IPC message is received there is no blink window bound to current scope, and Electron would crash if `blink::ToLocalDOMWindow` is called during JS executions.

By using `blink::WebLocalFrame::RequestExecuteV8Function` to execute JS functions, we can correctly setup a blink environment.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fix crash when debugging async/await code.

### CODE REVIEW

File: shell/renderer/electron_api_service_impl.cc [hunk 1]
```diff
@@ -54,24 +55,19 @@ void InvokeIpcCallback(v8::Local<v8::Context> context,
   if (ipcNative.IsEmpty())
     return;
 
-  // Only set up the node::CallbackScope if there's a node environment.
-  // Sandboxed renderers don't have a node environment.
-  node::Environment* env = node::Environment::GetCurrent(context);
-  std::unique_ptr<node::CallbackScope> callback_scope;
```
Comment (MarshallOfSound) [inline shell/renderer/electron_api_service_impl.cc]: I think we still need this when there is a node environment?

File: shell/renderer/electron_api_service_impl.cc [hunk 2]
```diff
@@ -54,24 +55,19 @@ void InvokeIpcCallback(v8::Local<v8::Context> context,
...
-
   auto callback_key = gin::ConvertToV8(isolate, callback_name)
                           ->ToString(context)
                           .ToLocalChecked();
   auto callback_value = ipcNative->Get(context, callback_key).ToLocalChecked();
   DCHECK(callback_value->IsFunction());  // set by init.ts
   auto callback = callback_value.As<v8::Function>();
-  ignore_result(callback->Call(context, ipcNative, args.size(), args.data()));
+
+  frame->RequestExecuteV8Function(context, callback, ipcNative, args.size(),
```
Comment (MarshallOfSound) [inline shell/renderer/electron_api_service_impl.cc]: This isn't guaranteed to be synchronous anymore correct? I don't think that's a problem just pointing it out.

File: shell/renderer/electron_api_service_impl.cc (hunk 1 above)
Comment (nornagon) [reply shell/renderer/electron_api_service_impl.cc]: I believe the node callback scope is responsible for invoking async hooks, which are totally broken in Electron currently, but shouldn't be 🙃

**Comment by deepak1556** [comment]
@zcbenz the problem seems to be somewhere with task runners being shared between chromium and node, here is a different code path.

In the test case from the issue replace the html file with following contents.

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Hello World!</title>
  </head>
  <body>
    <h1>Hello World!</h1>
  </body>
  <script>
	class Foo {
		constructor() {
			document.body.addEventListener('click', () => {
				this.foo();
			});
		}
		async foo() {
			require('fs').promises.readdir('electron-gn/src').then((err, files) => {
        console.log(files.length);
      });
			await this.foo2(); // set breakpoint in this line
			await this.foo2();
		}
		async foo2() {
			return Promise.resolve();
		}
	}
	new Foo();
  </script>
</html>
```


This doesn't go through the `EmitIPCEvent` code path on purpose, rather the promise resolution from node fs , check from frame 30 https://gist.github.com/deepak1556/2e60fb4c89134026fa27aaf845aabfb5 will trigger a microtask checkpoint and cause a similar crash.

File: shell/renderer/electron_api_service_impl.cc (hunk 1 above)
Comment (deepak1556) [reply shell/renderer/electron_api_service_impl.cc]: The problem is not with async hooks, we can bypass those checks with `export NODE_OPTIONS="--no-force-async-hooks-checks"`, although not a ideal flow.

Here the problem is about `node::CallbackScope` triggering a microtask checkpoint at an undesired time.

**Comment by zcbenz** [comment]
I'll try fixing it in another way.

**Pull request closed without merging**